flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java

```diff
@@ -100,21 +100,37 @@ protected CatalogLoader createCatalogLoader(String name, Map<String, String> pro
   }
 
   @Override
-  public Map<String, String> requiredContext() {
-    Map<String, String> context = Maps.newHashMap();
-    context.put(CatalogDescriptorValidator.CATALOG_TYPE, "iceberg");
-    context.put(CatalogDescriptorValidator.CATALOG_PROPERTY_VERSION, "1");
-    return context;
+  public String factoryIdentifier() {
+    return FlinkCatalogFactoryOptions.IDENTIFIER;
   }
 
   @Override
-  public List<String> supportedProperties() {
-    return ImmutableList.of("*");
+  public Set<ConfigOption<?>> requiredOptions() {
+    final Set<ConfigOption<?>> options = new HashSet<>();
+    options.add(FlinkCatalogFactoryOptions.CATALOG_TYPE);
+    return options;
   }
 
   @Override
-  public Catalog createCatalog(String name, Map<String, String> properties) {
-    return createCatalog(name, properties, clusterHadoopConf());
+  public Set<ConfigOption<?>> optionalOptions() {
+    final Set<ConfigOption<?>> options = new HashSet<>();
+    options.add(FlinkCatalogFactoryOptions.PROPERTY_VERSION);
+    options.add(FlinkCatalogFactoryOptions.URI);
+    options.add(FlinkCatalogFactoryOptions.WAREHOUSE);
+    options.add(FlinkCatalogFactoryOptions.CLIENTS);
+    options.add(FlinkCatalogFactoryOptions.BASE_NAMESPACE);
+    options.add(FlinkCatalogFactoryOptions.HIVE_CONF_DIF);
+    options.add(FlinkCatalogFactoryOptions.CACHE_ENABLED);
+    return options;
+  }
+
+  @Override
+  public Catalog createCatalog(Context context) {
+    final FactoryUtil.CatalogFactoryHelper helper =
+        FactoryUtil.createCatalogFactoryHelper(this, context);
+    helper.validate();
+
+    return createCatalog(context.getName(), context.getOptions(), clusterHadoopConf());
```

Comment:
There is an effort to decouple the need for Hadoop from the `FlinkCatalogFactory` for environments where Hadoop is not easily configurable and for catalog implementations that don't actually need it (basically anything that uses `S3FileIO` - though we'll need to update `GlueCatalog` and eventually `DynamoDbCatalog` as well).

Even after updating `GlueCatalog` to remove the Hadoop `Configurable` interface, this call to `clusterHadoopConf()` still makes it so that Hadoop is needed on the class path.

I'm not proposing that we change that in this PR (as this PR has been open for a while, and it's a separate concern), but I wanted to draw attention to the issue as I'm a bit less informed on the Flink side compared to many of you: https://github.com/apache/iceberg/issues/3044